methods_exploration/it_functions.py

```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Feb  8 10:11:52 2022
+
+@author: ggorski
+
+These functions were developed borrowing code and ideas from:
+    Laurel Larsen (laurel@berkeley.edu) for Larsen, L. G. and J. W. Harvey. 2017. Disrupted carbon cycling in restored and unrestored urban streams: Critical timescales and controls. Limnology and Oceanography, 62(Suppl. S1), S160-S182, doi: 10.1002/lno.10613. Please cite this work accordingly in products that use this code.
+    Edom Modges (edom.moges@berkeley.edu)
+    Julianne Quinn University of Virginia
+    The transfer entropy routine is based on Ruddell and Kumar, 2009, "Ecohydrologic process networks: 1. Identification," Water Resources Research, 45, doi:10.1029/2008WR007279.
+
+"""
+
+from joblib import Parallel, delayed
+import numpy as np
+
+
+def calc2Dpdf(M,nbins = 11):
+    '''calculates the 3 pdfs, one for x, one for y and a joint pdf for x and y 
+    M: a numpy array of shape (nobs, 2) where nobs is the number of observations
+    this assumes that the data are arrange such that the first column is the source and
+    the second column is the sink.
+    nbins: is the number of bins used for estimating the pdf with a default of 11'''
+    
+    counts, binEdges = np.histogramdd(M,bins=nbins)
+    p_xy = counts/np.sum(counts)
+    
+    p_x = np.sum(p_xy,axis=1)
+    p_y = np.sum(p_xy,axis=0)
+    
+    return p_x, p_y, p_xy
+
+
+
+def calcEntropy(pdf):
+    '''calculate the entropy from a the pdf'''
+    
+    pdf_n_0 = pdf[pdf>0]
+    log2_pdf_n_0 = np.log2(pdf_n_0)
+    H = (-sum(pdf_n_0*log2_pdf_n_0))
+    return H
+
+
+def calcMI(M, nbins = 11):
+    '''calculate mutual information of two variables
+    M: a numpy array of shape (nobs, 2) where nobs is the number of observations
+    this assumes that the data are arrange such that the first column is the source and
+    the second column is the sink.
+    nbins: is the number of bins used for estimating the pdf with a default of 11
```

Comment:
Why 11? It might help to make `nbins` a function of the length of `nobs`

Reply:
This what has been done in a few papers ([Tennant et al 2020](https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2019WR024908) and [Larsen and Harvey 2017](https://aslopubs.onlinelibrary.wiley.com/doi/10.1002/lno.10613)). Both of the papers use the codes that I am working off of. Their choices are mostly based on the analysis from[ Ruddell 2009 (see Appendix A1)](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2008WR007279). Ultimately, I think that having nbins be a variable (with guidance in the readme) makes the most sense, but I just set it here to 11 initially.